### PR TITLE
arrow-cpp: 3.0.0 -> 4.0.0

### PR DESCRIPTION
**Pull request by tobim**
###### Motivation for this change

The Arrow project decided to make [f959141 the new release](https://github.com/apache/arrow/releases/tag/apache-arrow-4.0.0): [1]

[1] https://lists.apache.org/thread.html/r4e590a68ff151b29b96780e249614a36f9a805911d82edf15c0a52f6%40%3Cdev.arrow.apache.org%3E

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 120854` at 45e32377 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python39Packages.google-cloud-bigquery</li>
  </ul>
</details>
<details>
  <summary>3 packages skipped due to time constraints:</summary>
  <ul>
    <li>python39Packages.datasets</li>
    <li>python39Packages.intake</li>
    <li>python39Packages.pyarrow</li>
  </ul>
</details>
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>arrow-cpp</li>
    <li>python38Packages.datasets</li>
    <li>python38Packages.google-cloud-bigquery</li>
    <li>python38Packages.intake</li>
    <li>python38Packages.pyarrow</li>
    <li>xsimd</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/development/libraries/arrow-cpp/default.nix:1:43:
```
  |
1 | { stdenv, lib, fetchurl, fetchFromGitHub, fetchpatch, fixDarwinDylibNames
  |                                           ^
```
</li>
  </ul>
</details>

File: pkgs/development/libraries/xsimd/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ lib, stdenv, fetchFromGitHub, cmake, gtest }:
+
+stdenv.mkDerivation rec {
+  pname = "xsimd";
+  version = "7.5.0";
+  src = fetchFromGitHub {
+    owner  = "xtensor-stack";
+    repo   = "xsimd";
+    rev    = "${version}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/xsimd/default.nix:9]:
```suggestion
  version = "7.5.0";

  src = fetchFromGitHub {
    owner = "xtensor-stack";
    repo = "xsimd";
    rev = version;
```

File: pkgs/development/libraries/xsimd/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  doCheck = true;
+  checkInputs = [ gtest ];
+  checkTarget = "xtest";
+
+  meta = with lib; {
+    description    = "C++ wrappers for SIMD intrinsics";
+    homepage       = "https://github.com/xtensor-stack/xsimd";
+    license        = licenses.bsd3;
+    maintainers    = with maintainers; [ tobim ];
+    platforms      = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/xsimd/default.nix:26]:
```suggestion
    description = "C++ wrappers for SIMD intrinsics";
    homepage = "https://github.com/xtensor-stack/xsimd";
    license = licenses.bsd3;
    maintainers = with maintainers; [ tobim ];
    platforms = platforms.all;
```

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 3]
```diff
@@ -111,6 +116,8 @@ in stdenv.mkDerivation rec {
     "-DCMAKE_INSTALL_RPATH=@loader_path/../lib" # needed for tools executables
   ] ++ lib.optional (!stdenv.isx86_64) "-DARROW_USE_SIMD=OFF";
 
+  ARROW_XSIMD_URL = "${xsimd.src}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/arrow-cpp/default.nix:119]:
```suggestion
  ARROW_XSIMD_URL = xsimd.src;
```

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 4]
```diff
@@ -1,6 +1,7 @@
 { stdenv, lib, fetchurl, fetchFromGitHub, fetchpatch, fixDarwinDylibNames
```
Comment (tobim) [inline pkgs/development/libraries/arrow-cpp/default.nix:1]:
```suggestion
{ stdenv, lib, fetchurl, fetchFromGitHub, fixDarwinDylibNames
```

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.pyarrow

**Comment by veprbl** [comment]
Fails on darwin:
```
/tmp/nix-build-xsimd-7.5.0.drv-0/source/include/xsimd/types/xsimd_scalar.hpp:412:9: error: use of undeclared identifier '__sincosf'; did you mea
n 'sincos'?
        __sincosf(val, &s, &c);
        ^~~~~~~~~
        sincos
/tmp/nix-build-xsimd-7.5.0.drv-0/source/include/xsimd/types/xsimd_scalar.hpp:409:17: note: 'sincos' declared here
    inline void sincos(float val, float&s, float& c)
                ^
/tmp/nix-build-xsimd-7.5.0.drv-0/source/include/xsimd/types/xsimd_scalar.hpp:412:24: error: non-const lvalue reference to type 'float' cannot bi
nd to a temporary of type 'float *'
        __sincosf(val, &s, &c);
                       ^~
/tmp/nix-build-xsimd-7.5.0.drv-0/source/include/xsimd/types/xsimd_scalar.hpp:409:41: note: passing argument to parameter 's' here
    inline void sincos(float val, float&s, float& c)
                                        ^
```

**Comment by tobim** [comment]
@veprbl would you like to take another look?

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 120854` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python39Packages.google-cloud-bigquery</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>arrow-cpp</li>
    <li>python38Packages.datasets</li>
    <li>python38Packages.google-cloud-bigquery</li>
    <li>python38Packages.intake</li>
    <li>python38Packages.pyarrow</li>
    <li>python39Packages.datasets</li>
    <li>python39Packages.intake</li>
    <li>python39Packages.pyarrow</li>
    <li>xsimd</li>
  </ul>
</details>

**Comment by veprbl** [comment]
I've submitted the issue upstream https://github.com/xtensor-stack/xsimd/pull/463

**Review by veprbl — APPROVED** [review]

File: pkgs/development/libraries/xsimd/default.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{ lib, stdenv, fetchgit, fetchFromGitHub, cmake, gtest }:
+let
+  version = "7.5.0";
+
+  darwin_src = fetchgit {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/xsimd/default.nix:5]: Please use fetchFromGitHub which is more performant as it can download tarballs.

**Comment by veprbl** [comment]
@tobim This is looking good. Would you mind squashing some commits?

**Comment by tobim** [comment]
> @tobim This is looking good. Would you mind squashing some commits?

Done.

**Comment by veprbl** [comment]
@tobim Thank you!

**Pull request merged**